### PR TITLE
Mm syslog sender

### PR DESCRIPTION
**Pull request by maimorag**
## Status
- [x] In Progress

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-4023

## Description
Adding support TLS option (currently UDP \ TCP)  ("Syslog" Pack) to  Syslog sender.

## Minimum version of Cortex XSOAR
- [X] 6.0.0

### CODE REVIEW

File: Packs/Syslog/Integrations/SyslogSender/SyslogSender.py [hunk 1]
```diff
@@ -116,13 +143,85 @@
...
+            # 1. add:
+            #   context.verify_mode = ssl.CERT_NONE
+            # 2. remove the lines:
+            #   context.load_verify_locations(certificate)
+            #   ssl_context.check_hostname = False
+            ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
+            ssl_context.check_hostname = False
+            ssl_context.verify_mode = ssl.CERT_NONE
+            ssl_context.load_verify_locations(self.certfile)
+            ssl_sock = ssl_context.wrap_socket(ssl_sock, server_hostname=self.address)
```
[the remark this answers is not in the document]
Comment (maimorag) [reply Packs/Syslog/Integrations/SyslogSender/SyslogSender.py]: not secure yet since I need to use self signed certificate.

File: Packs/Syslog/.secrets-ignore [hunk 2]
```diff
@@ -1,3 +1,5 @@
 SysLogHandler
 127.0.0.1
-www.eizelulz.com
\ No newline at end of file
+www.eizelulz.com
+aHR0cHM6Ly93d3cuZXhhbXBsZS5jb20vdGVzd
```
Comment (Shellyber) [inline Packs/Syslog/.secrets-ignore:4]: Are we sure that those are secrets??

File: Packs/Syslog/Integrations/SyslogSender/SyslogSender.py [hunk 3]
```diff
@@ -54,60 +61,129 @@
...
+            ssl_context.verify_mode = ssl.CERT_NONE
+
+        ssl_context.load_verify_locations(self.certfile)
+        ssl_sock = ssl_context.wrap_socket(ssl_sock, server_hostname=self.address)
+        self.socket = ssl_sock
+        try:
+            self.socket.connect((self.address, self.port))
+        except OSError as exc:
+            err = exc
+            if ssl_sock is not None:
```
Comment (Shellyber) [inline Packs/Syslog/Integrations/SyslogSender/SyslogSender.py:99]:
```suggestion
            if ssl_sock:
```

File: Packs/Syslog/Integrations/SyslogSender/SyslogSender.py [hunk 4]
```diff
@@ -54,60 +61,129 @@
...
+                ssl_sock.close()
+            if err is not None:
```
Comment (Shellyber) [inline Packs/Syslog/Integrations/SyslogSender/SyslogSender.py:101]:
```suggestion
            if err:
```

File: Packs/Syslog/Integrations/SyslogSender/SyslogSender.yml [hunk 5]
```diff
@@ -11,22 +11,35 @@ configuration:
   required: false
   type: 0
   section: Connect
-- defaultvalue: '514'
```
Comment (Shellyber) [inline Packs/Syslog/Integrations/SyslogSender/SyslogSender.yml:14]: Why did you remove the default?

**Review by Shellyber — COMMENTED** [review]
Great Work
Let's do a small demo

File: Packs/Syslog/Integrations/SyslogSender/SyslogSender.yml (hunk 5 above)
Comment (maimorag) [reply Packs/Syslog/Integrations/SyslogSender/SyslogSender.yml:14]: because when using TLS 6514 is the default port, so I added the default port while preparing the manager in the code- would you rather keep it anyway?

File: Packs/Syslog/.secrets-ignore (hunk 2 above)
Comment (maimorag) [reply Packs/Syslog/.secrets-ignore:4]: I got alart at start but now I don't so I'm removing it from there :)

File: Packs/Syslog/Integrations/SyslogSender/SyslogSender.yml [hunk 6]
```diff
@@ -11,22 +11,35 @@ configuration:
...
-  display: Port
+  additionalinfo: A Syslog server address must be provided.
+- display: Port
   name: port
   required: false
   type: 0
   section: Connect
+  additionalinfo: When using UDP or TCP, port 514 is the default. Else,when using TLS, port 6514 is the default.
```
Comment (ShirleyDenkberg) [inline Packs/Syslog/Integrations/SyslogSender/SyslogSender.yml:20]:
```suggestion
  additionalinfo: When using UDP or TCP, port 514 is the default. When using TLS, port 6514 is the default.
```

File: Packs/Syslog/Integrations/SyslogSender/SyslogSender.yml [hunk 7]
```diff
@@ -172,7 +183,7 @@ script:
       secret: false
     - auto: PREDEFINED
       default: false
-      description: The protocol to use. Can be "TCP" or "UDP".
+      description: The protocol to use. Can be "TCP" or "UDP" or "TLS".
```
Comment (ShirleyDenkberg) [inline Packs/Syslog/Integrations/SyslogSender/SyslogSender.yml:186]:
```suggestion
      description: The protocol to use. Can be "TCP", "UDP", or "TLS".
```

File: Packs/Syslog/ReleaseNotes/2_0_13.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+
+#### Integrations
+##### Syslog Sender
+- Added a support in TLS.
```
Comment (ShirleyDenkberg) [inline Packs/Syslog/ReleaseNotes/2_0_13.md:4]:
```suggestion
- Added support for TLS.
```

**Comment by ShirleyDenkberg** [comment]
@Shellyber Doc review completed.

File: Packs/Syslog/Integrations/SyslogSender/SyslogSender.py [hunk 9]
```diff
@@ -54,60 +61,135 @@
...
     @contextmanager  # type: ignore[misc, arg-type]
     def get_logger(self) -> Generator:
         """
         Get a new instance of a syslog logger.
         :return: syslog logger
         """
-        handler = self._get_handler()
+        if self.protocol == TLS and self.syslog_cert_path:
+            handler = self.init_handler_tls(self.syslog_cert_path)
+            demisto.debug('get handler tls')
```
Comment (DeanArbel) [inline Packs/Syslog/Integrations/SyslogSender/SyslogSender.py:165]: Can this init fail? if so it's a better idea to print the log before it.
One more thing - I don't understand what "get handler tls" means. Perhaps `creating tls logger handler`

File: Packs/Syslog/Integrations/SyslogSender/SyslogSender.py [hunk 10]
```diff
@@ -54,60 +61,135 @@
...
-
-        return SysLogHandler(**kwargs)
-
-    def _init_logger(self, handler: SysLogHandler) -> Logger:
+    def _get_handler(self) -> Rfc5424SysLogHandler:
+        sock_kind = SOCK_STREAM if self.protocol == TCP else socket.SOCK_DGRAM
+        return Rfc5424SysLogHandler(address=(self.address, self.port),
+                                    facility=self.facility,
+                                    socktype=sock_kind,
+                                    timeout=10,
```
Comment (DeanArbel) [inline Packs/Syslog/Integrations/SyslogSender/SyslogSender.py:181]: Why 10?
What's the default timeout if you leave it empty?

File: Packs/Syslog/Integrations/SyslogSender/SyslogSender.py [hunk 11]
```diff
@@ -116,31 +198,58 @@ def _init_logger(self, handler: SysLogHandler) -> Logger:
...
     port = int(params.get('port', 514))
     protocol = params.get('protocol', UDP).lower()
     facility = FACILITY_DICT.get(params.get('facility', 'LOG_SYSLOG'), SysLogHandler.LOG_SYSLOG)
     logging_level = LOGGING_LEVEL_DICT.get(params.get('priority', 'LOG_INFO'), INFO)
-
+    certificate: Optional[str] = (replace_spaces_in_credential(params.get('certificate', {}).get('password'))
+                                  or params.get('certificate', None))
+    certificate_path: Optional[str] = None
+    max_port = 65535
+    default_port = 6514 if protocol == 'tls' else 514
```
Comment (DeanArbel) [inline Packs/Syslog/Integrations/SyslogSender/SyslogSender.py:238]: Please extract these values to a constant, i.e.
```
MAX_PORT = 65535
DEFAULT_TCP_SYSLOG_PORT = 514
DEFAULT_TLS_SYSLOG_PORT = 6514
```

File: Packs/Syslog/Integrations/SyslogSender/SyslogSender.py [hunk 12]
```diff
@@ -116,31 +198,58 @@ def _init_logger(self, handler: SysLogHandler) -> Logger:
...
-        raise ValueError('A Syslog server address must be provided.')
-    if not port and protocol in PROTOCOLS:
-        raise ValueError('A port must be provided in TCP or UDP protocols.')
-
-    return SyslogManager(address, port, protocol, logging_level, facility)
+        raise DemistoException('A address must be provided.')
+    try:
+        port = int(params.get('port', default_port))
+    except (ValueError, TypeError):
+        raise DemistoException(f'Invalid listen port - {port}. Make sure your port is a number')
```
Comment (DeanArbel) [inline Packs/Syslog/Integrations/SyslogSender/SyslogSender.py:245]: you can replace this with `arg_to_number` from CommonServerPython.

**Review by DeanArbel — CHANGES_REQUESTED** [review]
Looks good, just a few comments.

File: Packs/Syslog/Integrations/SyslogSender/SyslogSender.py (hunk 10 above)
Comment (maimorag) [reply Packs/Syslog/Integrations/SyslogSender/SyslogSender.py:181]: 5, should I remove it then?

Comment (maimorag) [reply Packs/Syslog/Integrations/SyslogSender/SyslogSender.py]: I removed it

**Review by Shellyber — APPROVED** [review]
Great work!
@maimorag Please fix the pre-commit step before merging

**Review by DeanArbel — APPROVED** [review]
Looks good
